### PR TITLE
add error code for rest api

### PR DESCRIPTION
**Pull request by zccst**
______
* distrbution
https://github.com/okex/okexchain/pull/515/commits/bba3505fd8470bf6a69333936023fc402f54e44e
*ammswap
https://github.com/okex/okexchain/pull/515/commits/0f8fa3d28fe6718fb7653cd6765e8e0540908c61
* dex
https://github.com/okex/okexchain/pull/515/commits/2150a2e93e0e3bcb0e5796464790bede27fe56cc
https://github.com/okex/okexchain/pull/515/commits/070db257fdc5b906b546b553f51ff34bba2c753e
https://github.com/okex/okexchain/pull/515/commits/920581a1189732d5f9a686c8c890993d458a77b0
https://github.com/okex/okexchain/pull/515/commits/6a3017d7e31d82fa1ca6dccbd96e72fcce29f29d
_______
Modules approved
- [x] ammswap         
- [x] common          
- [x] dex             
- [x] evidence        
- [x] farm
- [x] gov             
- [x] params          
- [x] staking         
- [x] token
- [x] backend         
- [x] debug           
- [x] distribution    
- [x] evm             
- [x] genutil         
- [x] order           
- [x] slashing        
- [x] stream

### CODE REVIEW

File: x/token/token_test.go [hunk 1]
```diff
@@ -20,21 +14,25 @@ import (
 	"github.com/okex/okexchain/x/token/types"
 	"github.com/stretchr/testify/require"
 	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/crypto"
 	"github.com/tendermint/tendermint/crypto/secp256k1"
+	"strconv"
```
Comment (ylsGit) [inline x/token/token_test.go]: import order

File: x/token/types/msgs.go [hunk 2]
```diff
@@ -43,33 +44,33 @@ func (msg MsgTokenIssue) Type() string { return "issue" }
...
 	}
 
 	// check original symbol
 	if len(msg.OriginalSymbol) == 0 {
-		return sdk.ErrUnknownRequest("failed to check issue msg because original symbol cannot be empty")
+		return ErrUserInputSymbolIsEmpty()
 	}
 	if !ValidOriginalSymbol(msg.OriginalSymbol) {
-		return sdk.ErrUnknownRequest("failed to check issue msg because invalid original symbol: " + msg.OriginalSymbol)
+		return ErrNotAllowedOriginalSymbol()
```
Comment (ylsGit) [inline x/token/types/msgs.go]: missing msg.OriginalSymbol

File: x/token/types/msgs.go [hunk 3]
```diff
@@ -43,33 +44,33 @@ func (msg MsgTokenIssue) Type() string { return "issue" }
 func (msg MsgTokenIssue) ValidateBasic() sdk.Error {
 	// check owner
 	if msg.Owner.Empty() {
-		return sdk.ErrInvalidAddress(msg.Owner.String())
+		return ErrAddressIsRequired()
```
Comment (ylsGit) [inline x/token/types/msgs.go:47]: missing params

File: x/token/handler.go [hunk 4]
```diff
@@ -105,9 +104,7 @@ func handleMsgTokenIssue(ctx sdk.Context, keeper Keeper, msg types.MsgTokenIssue
 	// generate a random symbol
 	newName, valid := addTokenSuffix(ctx, keeper, msg.OriginalSymbol)
 	if !valid {
-		return sdk.ErrInvalidCoins(fmt.Sprintf(
-			"temporarily failed to generate a unique symbol for %s. Try again.",
-			msg.OriginalSymbol)).Result()
+		return types.ErrInvalidCoins().Result()
```
Comment (ylsGit) [inline x/token/handler.go]: missing params

File: x/staking/types/errors.go [hunk 5]
```diff
@@ -12,264 +12,303 @@ import (
...
-// ErrNoValidatorFound returns an error when a validator doesn't exist
-func ErrNoValidatorFound(codespace string, valAddr string) sdk.EnvelopedErr {
-	return sdk.EnvelopedErr{Err: sdkerrors.New(codespace, CodeInvalidValidator, fmt.Sprintf("validator %s does not exist", valAddr))}
+// ErrMoreMinSelfDelegation returns an error when the msd doesn't match the rest of shares on a validator
+func ErrMoreMinSelfDelegation(valAddr string) sdk.Error {
+	return sdkerrors.New(DefaultCodespace, CodeMoreMinSelfDelegation,
+		fmt.Sprintf("failed. min self delegation of %s is more than its shares", valAddr))
+}
+
+// ErrProxyNotFound returns an error when a delegator who's not a proxy trys to unreg
```
Comment (RevelationOfTuring) [inline x/staking/types/errors.go]: trys -> tries

File: x/staking/types/errors.go [hunk 6]
```diff
@@ -12,264 +12,303 @@ import (
...
+		fmt.Sprintf("failed. proxy %s not found", delegator))}
+}
+
+// ErrEmptyValidators returns an error when none of the validators in shares adding list is available
+func ErrEmptyValidators() sdk.Error {
+	return sdkerrors.New(DefaultCodespace, CodeEmptyValidators,
+		"failed. empty validators")
+}
+
+// ErrProxyAlreadyExist returns an error when a proxy trys to reg the second time
```
Comment (RevelationOfTuring) [inline x/staking/types/errors.go]: trys -> tries

File: x/staking/types/errors.go [hunk 7]
```diff
@@ -12,264 +12,303 @@ import (
...
+func ErrNilDelegatorAddr() sdk.Error {
+	return sdkerrors.New(DefaultCodespace, CodeNilDelegatorAddr, "delegator address is nil")
 }
 
-// ErrWrongOperationAddr returns an error when the address is not expected
-func ErrWrongOperationAddr(codespace string, msg string) sdk.Error {
-	return sdkerrors.New(codespace, CodeInvalidInput, fmt.Sprintf("wrong operation address found. msg: %s", msg))
+// ErrDelegatorEqualToProxyAddr returns an error when the address is not expected
+func ErrDelegatorEqualToProxyAddr() sdk.Error {
+	return sdkerrors.New(DefaultCodespace, CodeDelegatorEqualToProxyAddr, "delegator address can not eqauls to proxy address")
```
Comment (RevelationOfTuring) [inline x/staking/types/errors.go]: eqauls -> equals

File: x/staking/types/errors.go [hunk 8]
```diff
@@ -12,264 +12,303 @@ import (
...
-	return sdk.EnvelopedErr{Err: sdkerrors.New(codespace, CodeInvalidDelegation, "invalid coin denomination")}
+func ErrBadDenom() sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeBadDenom, "invalid coin denomination")}
 }
 
 // ErrBadDelegationAmount returns an error when the amount of delegation isn't positive
-func ErrBadDelegationAmount(codespace string) sdk.Error {
-	return sdkerrors.New(codespace, CodeInvalidDelegation, "amount must be greater than 0")
+func ErrBadDelegationAmount() sdk.Error {
+	return sdkerrors.New(DefaultCodespace, CodeBadDelegationAmount, "amount must more than 0")
```
Comment (RevelationOfTuring) [inline x/staking/types/errors.go]: must be more than ...

**Review by banishee — APPROVED** [review]
x/distr is approved

File: x/backend/types/errors.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+func ErrProductDoesNotExist(product string) sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeProductDoesNotExist, fmt.Sprintf("product %s does not exist", product))}
+}
+
+func ErrBackendPluginNotEnabled() sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeBackendPluginNotEnabled, "backend is not enabled")}
+}
+
+func ErrParamNotCorrect(size int, granularity int) sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeParamNotCorrect, fmt.Sprintf("parameter's not correct, size: %d, granularity: %d", size, granularity))}
```
Comment (xiangjianmeng) [inline x/backend/types/errors.go]: parameter is rather than parameter's

File: x/common/errors.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeMarshalJSONFailed, fmt.Sprintf("could not marshal result to JSON, %s", msg))}
+}
+
+// could not unmarshal result to origin
+func ErrUnMarshalJSONFailed(msg string) sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeUnMarshalJSONFailed, fmt.Sprintf("incorrectly formatted request data, %s", msg))}
+}
+
+func ErrStrconvFailed(msg string) sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeStrconvFailed, fmt.Sprintf("incorrectly string conversion"))}
```
Comment (xiangjianmeng) [inline x/common/errors.go]: suggest to use "incorrect string conversion"

**Review by ylsGit — APPROVED** [review]
x/token approved

**Review by RevelationOfTuring — APPROVED** [review]
staking approve

File: x/order/types/errors.go [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// const uint32
+const (
+	CodeInvalidAddress                        uint32 = 63000
+	CodeSizeIsInvalid                         uint32 = 63001
+	CodeTokenPairNotExist                     uint32 = 63002
+	CodeSendCoinsFaile                        uint32 = 63003
```
Comment (summerpro) [inline x/order/types/errors.go]: should be CodeSendCoinsFailed

**Review by xiangjianmeng — COMMENTED** [review]
backend approve

**Review by xiangjianmeng — COMMENTED** [review]
common approve

**Comment by banishee** [comment]
ammswap is approved

File: x/params/types/proposal.go [hunk 12]
```diff
@@ -38,24 +38,22 @@ func NewParameterChangeProposal(title, description string, changes []types.Param
 // ValidateBasic validates the parameter change proposal
 func (pcp ParameterChangeProposal) ValidateBasic() sdk.Error {
 	if len(strings.TrimSpace(pcp.Title)) == 0 {
-		return govtypes.ErrInvalidProposalContent(DefaultCodespace, "proposal title cannot be blank")
+		return govtypes.ErrInvalidProposalContent("title is required")
 	}
 	if len(pcp.Title) > govtypes.MaxTitleLength {
-		return govtypes.ErrInvalidProposalContent(DefaultCodespace,
-			fmt.Sprintf("proposal title is longer than max length of %d", govtypes.MaxTitleLength))
+		return govtypes.ErrInvalidProposalContent("title length is bigger than max title length")
```
Comment (ilovers) [inline x/params/types/proposal.go]: replace "bigger" to "longer" .

File: x/params/types/proposal.go [hunk 13]
```diff
@@ -38,24 +38,22 @@ func NewParameterChangeProposal(title, description string, changes []types.Param
...
 
 	if len(pcp.Description) == 0 {
-		return govtypes.ErrInvalidProposalContent(DefaultCodespace, "proposal description cannot be blank")
+		return govtypes.ErrInvalidProposalContent("description is required")
 	}
 
 	if len(pcp.Description) > govtypes.MaxDescriptionLength {
-		return govtypes.ErrInvalidProposalContent(DefaultCodespace,
-			fmt.Sprintf("proposal description is longer than max length of %d", govtypes.MaxDescriptionLength))
+		return govtypes.ErrInvalidProposalContent("description length is bigger than max DggescriptionLength")
```
Comment (ilovers) [inline x/params/types/proposal.go]: replace "bigger" to "longer" .

File: x/dex/keeper/proposal.go [hunk 14]
```diff
@@ -39,32 +39,32 @@ func (k Keeper) GetVotingPeriod(ctx sdk.Context, content gov.Content) (votingPer
...
+		return types.ErrTokenPairNotFound(fmt.Sprintf("%s_%s", delistProposal.BaseAsset, delistProposal.QuoteAsset))
 	}
 
 	// check the initial deposit
 	localMinDeposit := k.GetParams(ctx).DelistMinDeposit.MulDec(sdk.NewDecWithPrec(1, 1))
 	err := common.HasSufficientCoins(proposer, initialDeposit, localMinDeposit)
 
 	if err != nil {
-		return types.ErrInvalidAsset(fmt.Sprintf("failed to submit proposal because initial deposit should be more than %s", localMinDeposit.String()))
+		return types.ErrInvalidAsset(fmt.Sprintf(localMinDeposit.String()))
```
Comment (RevelationOfTuring) [inline x/dex/keeper/proposal.go]: remove fmt.Sprintf()

**Review by ilovers — COMMENTED** [review]
x/params approved
x/stream approved

**Review by ilovers — APPROVED** [review]

File: x/order/types/errors.go [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+func ErrSendCoinsFaile(coins string, to string) sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeSendCoinsFaile, fmt.Sprintf("send fee(%s) to address(%s) failed\n", coins, to))}
+}
+
+func ErrTradingPairIsDelisting(product string) sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeTradingPairIsDelisting, fmt.Sprintf("trading pair '%s' is delisting", product))}
+}
+
+func ErrPriceOverAccuracy(price sdk.Dec, priceDigit int64) sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeRoundedPriceEqual, fmt.Sprintf("price(%v) over accuracy(%d)", price, priceDigit))}
```
Comment (summerpro) [inline x/order/types/errors.go]: CodeRoundedPriceEqual -> CodePriceOverAccuracy

CodeRoundedQuantityEqual -> CodeQuantityOverAccuracy

File: x/order/types/errors.go [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+
+func ErrOrderStatusIsNotOpen() sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeOrderStatusIsNotOpen, "order status is not open")}
+}
+
+func ErrOrderIsNotExistOrClosed(orderID string) sdk.EnvelopedErr {
+	return sdk.EnvelopedErr{Err: sdkerrors.New(DefaultCodespace, CodeOrderIsNotExistOrClosed, fmt.Sprintf("order(%s) does not exist or already closed", orderID))}
+}
+
+func ErrUnknowOrdernQueryType() sdk.EnvelopedErr {
```
Comment (summerpro) [inline x/order/types/errors.go]: ErrUnknownOrderQueryType

**Comment by summerpro** [comment]
order approved

**Comment by banishee** [comment]
gov is approved

**Review by RevelationOfTuring — APPROVED** [review]
dex : approved

**Comment by ilovers** [comment]
x/farm approved

**Comment by summerpro** [comment]
ammswap approve

**Pull request merged**